### PR TITLE
feat: Migrate `createSchema()` to TypeScript

### PR DESCRIPTION
**Pull request by evanrittenhouse**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [x] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

Migrate `createSchema()` to TypeScript. As an added bonus, this PR creates an initial implementation of `AbstractQueryInterfaceTypeScript` with `createSchema()`.

### CODE REVIEW

File: src/dialects/abstract/abstract-query-interface.ts
Comment (ephys) [inline src/dialects/abstract/abstract-query-interface.ts:5]: We don't use default exports in the project to maximise compat between esm & cjs

```suggestion
export class AbstractQueryInterface implements QueryInterface {
```

**Review by ephys — CHANGES_REQUESTED** [review]
We already started doing something like this but it's done the other way around. Take a look at `query-generator-typescript`, `model-typescript`, and `sequelize-typescript`

The class exported by `query-interface.js` is the one that should be renamed to `AbstractQueryInterface` (without changing the filename)

This TypeScript file should be named `query-interface-typescript.ts` and export `AbstractQueryInterfaceTypeScript`, `AbstractQueryInterface` should extend `AbstractQueryInterfaceTypeScript` in both `query-interface.d.ts` and `query-interface.js`

**Comment by WikiRik** [comment]
I quite like the idea of removing the exported interfaces from the class file but if we merge it here, we should also do this consistently for the other files (when we migrate those to TS files). This includes interfaces like `TransactionOptions` in `transaction.ts`
https://github.com/sequelize/sequelize/blob/bd037c80871a7bfff178887d4f0a6ae5c30a1e99/src/transaction.ts#L412-L425
but also `QueryGeneratorOptions` in `dialects/abstract/query-generator-typescript.ts`
https://github.com/sequelize/sequelize/blob/bd037c80871a7bfff178887d4f0a6ae5c30a1e99/src/dialects/abstract/query-generator-typescript.ts#L13-L16

And maybe we should do the same structure for exported types? Or would that create too many different files? (and even more extreme, the various functions in a file like `utils/model-utils.ts`. Are those files small enough to warrant multiple functions or do we want to separate every export to a different file)


This discussion is not really directly related to this PR, but might be good to discuss

**Comment by ephys** [comment]
I'm actually very conflicted by the idea of moving each interface to its individual file, having a ton of files is just as hard to navigate as having a single file. I think it's important to group by theme.

If we want to split them, I'd much prefer to have a single `query-interface-types.ts` that exports the different types that are currently declared in `query-interface.d.ts` over having 60 files that are 4-15 lines long

**Comment by evanrittenhouse** [comment]
I'm of the opinion that the class file should contain only logic associated with the class. At the same time I agree that it can get messy with tons of short interface files around.

Overall though, I think we should optimize for consistency, which is why I separated the types into separate files - if we have, say, 60 types for `query-interface` that are 4-15 lines each, but one that's 500 (let's say), for consistency's sake we'd put them all in one file that's 700 lines. This combines logic and responsibilities that may not need to be together. In personal projects I generally opt for splitting them all out into files under one directory which is why I went for this approach here. I would generally make `/query-interface/QueryInterface.ts` for the class logic, then `query-interface/types/CollateCharsetOptions.ts` for example. This falls apart when types should be re-used across several packages (`query-interface/` here is what I'm referring to as a package), but I think that points to a larger flaw in codebase design at that point. 

Just explaining my thoughts though, and happy to go in another direction! The TS migration is a huge opportunity to remove tech debt in the library, so I think it's important that we think about everything from implementation to organization. It's a good discussion to have.

**Comment by WikiRik** [comment]
Okay, so I think for this PR we should keep the types together with the classes and we can do a further (more in-depth) discussion in a different issue (possibly in the meetings repo) to refactor this in a later stage of v7. Since I do not think that the TypeScript files will be too long at the first release of v7 and as discussed at the monthly we would like to release a beta for v7 soon.

**Comment by evanrittenhouse** [comment]
@WikiRik Sounds good, though to be clear in my mind this PR is closer to [this comment[(https://github.com/sequelize/sequelize/issues/15260#issuecomment-1327572831), where we want to get `QueryGenerator` to a place where it can act as a public API. The code is more to just set a code baseline for how we'll do things, there's not really any logic change. Ideally I'd like to get this PR to a point where we have `QueryGenerator#experimental_select_query` done in a TS-centric manner before merging. Two questions: 
1. Does that make sense?
2. Will this hold up v7?

**Comment by WikiRik** [comment]
Ah, in that case you should check out this comment by @ephys ; https://github.com/sequelize/sequelize/pull/15339#pullrequestreview-1194967938

The aim of the temporary QueryGenerator (and QueryInterface) TS files is to only have methods that are suitable for the public API. But we have not yet determined how we will deal with the methods that we want to keep internal (we were thinking about an `InternalQueryInterfaceTypeScript` I believe)

File: src/dialects/abstract/abstract-query-interface.ts
Comment (evanrittenhouse) [reply src/dialects/abstract/abstract-query-interface.ts:5]: Gotcha - what's the difference between this TS file and `query-interface.d.ts`? Would `query-interface.d.ts` contain the supporting types and `AbstractQueryInterfaceTypeScript` would contain the actual interface definition?

File: src/dialects/abstract/query-interface.d.ts
Comment (ephys) [inline src/dialects/abstract/query-interface.d.ts:284]: don´t forget to `extends AbstractQueryInterfaceTypescript` in `query-interface.js` too

File: src/dialects/abstract/query-interface-typescript.ts
Comment (ephys) [inline src/dialects/abstract/query-interface-typescript.ts:6]:
```suggestion
  queryGenerator: AbstractQueryGenerator;
```

File: src/dialects/abstract/query-interface-typescript.ts
Comment (ephys) [inline src/dialects/abstract/query-interface-typescript.ts:11]:
```suggestion
  protected readonly queryGenerator: AbstractQueryGenerator;
```

File: src/dialects/abstract/query-interface.d.ts
Comment (evanrittenhouse) [reply src/dialects/abstract/query-interface.d.ts:284]: Yep, should've marked this as a draft PR!

**Comment by WikiRik** [comment]
Not sure why but the base branch was set to `query_interface_public_api` while this PR needs to land in `main`

**Comment by evanrittenhouse** [comment]
Marking as ready for review after only implementing `createSchema` since I was having some difficulties and want to ensure I'm going about it the right way, in terms of both implementation and typings.

File: src/dialects/abstract/query-interface-typescript.ts
Comment (ephys) [inline src/dialects/abstract/query-interface-typescript.ts:28]: I don't think you need to constantly check whether a parameter has been received, that's already what we use typescript for

File: src/dialects/abstract/query-interface-typescript.ts
Comment (ephys) [inline src/dialects/abstract/query-interface-typescript.ts:10]: These two should remain public for now. It's the only way to access the sequelize instance in migrations (until we rework migrations), and the only way to access queryGenerator altogether (until we expose it on the sequelize instance)

File: src/dialects/abstract/query-interface-typescript.ts
Comment (ephys) [inline src/dialects/abstract/query-interface-typescript.ts:19]: same here about checking being redundant with typescript

File: src/dialects/abstract/query-interface.d.ts
Comment (ephys) [inline src/dialects/abstract/query-interface.d.ts:284]: don't forget to remove create schema from this file, since it's now in the parent class

File: src/dialects/abstract/query-interface.d.ts
Comment (ephys) [inline src/dialects/abstract/query-interface.d.ts:323]: `ListSchemasQueryOptions` is only the options accepted by queryGenerator, but the options are also passed to queryRaw. This interface should accept both

File: src/dialects/abstract/query-interface.js
Comment (ephys) [inline src/dialects/abstract/query-interface.js:19]: redundant comment

File: src/dialects/abstract/query-interface.js
Comment (ephys) [inline src/dialects/abstract/query-interface.js:20]: Could you rename this one to AbstractQueryInterface?

Comment (ephys) [inline src/dialects/abstract/query-interface.js:20]: don't forget to remove create schema from this file as well

File: test/integration/query-interface/schemas.test.ts
Comment (ephys) [inline test/integration/query-interface/schemas.test.ts:7]: we call this `dialectName` by convention. `dialect` is always the dialect instance (`sequelize.dialect`)

File: test/integration/query-interface/schemas.test.ts
Comment (ephys) [inline test/integration/query-interface/schemas.test.ts:9]: should run on all dialects that support schemas (`dialect.supports.schemas`)

File: test/integration/query-interface/schemas.test.ts
Comment (ephys) [inline test/integration/query-interface/schemas.test.ts:14]: this expect should run before createSchema, as well as this change:

```suggestion
      expect(preCreationSchemas).to.not.include(testSchema, 'testSchema already exists prior to creation');
```

File: test/integration/query-interface/schemas.test.ts
Comment (ephys) [inline test/integration/query-interface/schemas.test.ts:17]: you didn't drop the schema but the test states that it's testing against that too

Comment (ephys) [inline test/integration/query-interface/schemas.test.ts:17]:
```suggestion
      expect(postCreationSchemas).to.include(testSchema, 'createSchema did not create testSchema');
```

**Review by ephys — CHANGES_REQUESTED** [review]

File: src/dialects/abstract/query-interface-typescript.ts
Comment (WikiRik) [reply src/dialects/abstract/query-interface-typescript.ts:19]: We are doing this in the `AbstractQueryGeneratorTypeScript` as well

File: src/dialects/abstract/query-interface-typescript.ts
Comment (WikiRik) [inline src/dialects/abstract/query-interface-typescript.ts:25]: Probably out of scope for this PR but;
We have `rejectInvalidOptions` implemented for `createSchemaQuery`. Why don't we pick the options that are in `CreateSchemaQueryOptions` and pass them on? (so not all options which are part of `QueryRawOptions` but only `collate` and `charset` for now)

File: src/dialects/abstract/query-interface-typescript.ts
Comment (WikiRik) [inline src/dialects/abstract/query-interface-typescript.ts:24]: I think we do need to extend these with `CreateSchemaQueryOptions`, right? Or are they already included in `QueryRawOptions`?

File: src/dialects/abstract/query-interface-typescript.ts
Comment (WikiRik) [inline src/dialects/abstract/query-interface-typescript.ts:31]: Don't forget to pass options to here as well

File: src/dialects/abstract/query-interface-typescript.ts
Comment (WikiRik) [inline src/dialects/abstract/query-interface-typescript.ts:29]: It might be good to create a `CreateSchemaOptions` interface that extends both `CreateSchemaQueryOptions` and `QueryRawOptions`

File: src/sequelize.d.ts
Comment (WikiRik) [inline src/sequelize.d.ts:938]: Here the same comment about `CreateSchemaOptions`

File: src/sequelize.js
Comment (WikiRik) [inline src/sequelize.js:787]: Maybe we should move this to `sequelize-typescript.ts`? If we want to keep supporting this after QueryInterface is public. cc @ephys

File: test/integration/query-interface/schemas.test.ts
Comment (WikiRik) [inline test/integration/query-interface/schemas.test.ts:27]:
```suggestion
      const postCreationSchemas = await queryInterface.showAllSchemas();
```
TypeScript should already know that the return of `showAllSchemas` is `string[]`, right?

File: test/integration/query-interface/schemas.test.ts
Comment (WikiRik) [inline test/integration/query-interface/schemas.test.ts:16]:
```suggestion
      const schemaNames = await queryInterface.showAllSchemas();
```
See comment below

File: test/integration/query-interface/schemas.test.ts
Comment (WikiRik) [inline test/integration/query-interface/schemas.test.ts:31]:
```suggestion
    it(`should pass options through to the queryInterface's queryGenerator`, async () => {
```

File: src/sequelize.js
Comment (evanrittenhouse) [reply src/sequelize.js:787]: Will implement and resolve after feedback from @ephys

File: src/dialects/abstract/query-interface-typescript.ts
Comment (evanrittenhouse) [reply src/dialects/abstract/query-interface-typescript.ts:29]: Was thinking about doing that but held off on implementing until after we had that talk about where we want to put our interfaces vs. class logic. I'll create it here and then we can move it later.

File: src/sequelize.js
Comment (evanrittenhouse) [reply src/sequelize.js:787]: Though  big gripe I've seen from the community is too many ways to do things - may be worth suppressing this method and keeping it in QueryInterface

File: test/integration/query-interface/schemas.test.ts
Comment (evanrittenhouse) [reply test/integration/query-interface/schemas.test.ts:27]: Yeah, I just like explicit typing. Will remove

File: src/dialects/abstract/query-interface-typescript.ts
Comment (ephys) [inline src/dialects/abstract/query-interface-typescript.ts:22]: I think we can remove this `@see`

File: src/dialects/abstract/query-interface-typescript.ts
Comment (ephys) [inline src/dialects/abstract/query-interface-typescript.ts:15]: So far our jsdoc has described methods in the third person

```suggestion
   * Creates a new database schema.
```

File: src/sequelize.js
Comment (ephys) [reply src/sequelize.js:787]: I agree, creating schemas isn't something we do every 5 minutes, there is no reason to have a shortcut method.

Plus, migrations already receive queryInterface instead of the sequelize instance

Model also had a createSchema method for some reason, which I think we removed in v7

I'm pro deprecating this method (doesn't have to be in this PR)

I'm also pro-deprecating `getQueryInterface`, it's redundant with `queryInterface`

File: test/integration/query-interface/schemas.test.ts
Comment (ephys) [inline test/integration/query-interface/schemas.test.ts:12]:
```suggestion
  describe('QueryInterface#createSchema', async () => {
```

File: test/integration/query-interface/schemas.test.ts
Comment (ephys) [inline test/integration/query-interface/schemas.test.ts:15]: This is already what the global afterEach does (unless `disableDatabaseResetForSuite` is called in the suite), so you don't need to have this :)

File: test/integration/query-interface/schemas.test.ts
Comment (ephys) [inline test/integration/query-interface/schemas.test.ts:22]: We don't follow the should convention because it's needlessly verbose

```suggestion
    it('creates a schema', async () => {
```

File: test/integration/query-interface/schemas.test.ts
Comment (ephys) [inline test/integration/query-interface/schemas.test.ts:41]: this expect should be after the try-catch block, it should run even if queryGenerator threw an error

File: test/integration/query-interface/schemas.test.ts
Comment (ephys) [inline test/integration/query-interface/schemas.test.ts:43]: this line is unnecessary

**Review by ephys — CHANGES_REQUESTED** [review]

File: src/dialects/abstract/query-interface.types.ts
Comment (ephys) [inline src/dialects/abstract/query-interface.types.ts:9]: I think these types could be documented with something like this for our API reference website:

```
/** Options accepted by {@link AbstractQueryInterfaceTypeScript#createSchema} */
```

File: test/integration/query-interface/schemas.test.ts
Comment (WikiRik) [inline test/integration/query-interface/schemas.test.ts:12]: Can we avoid nesting by using `if (!sequelize.dialect.supports.schemas)` here and return directly?

**Review by WikiRik — COMMENTED** [review]
One final comment and then we should be good to go

File: test/integration/query-interface/schemas.test.ts
Comment (evanrittenhouse) [reply test/integration/query-interface/schemas.test.ts]: Done - amended the latest commit and force-pushed due to how small the change is @WikiRik

**Review by WikiRik — APPROVED** [review]

**Review by ephys — APPROVED** [review]

File: src/sequelize.js
Comment (ephys) [inline src/sequelize.js:1283]: Last minute bug:

The type export doesn't match the export.
The export is still QueryInterface but it's now typed as AbstractQueryInterface

Comment (evanrittenhouse) [reply src/sequelize.js:1283]: @ephys What should it be? Does it make sense to just change the underlying type without changing the export?

Comment (ephys) [reply src/sequelize.js:1283]: We can have it exported as a different name than what we use internally, but the type export & the value export must have the same name

I'd go with QueryInterface outside of Sequelize for now to avoid a breaking change, which can be done by changing how index.d.ts exports QueryInterface

On top of that, the types added to query-interface.types.ts should be public too, need to export them from index.d.ts

Comment (evanrittenhouse) [reply src/sequelize.js:1283]: Implementing this in #15441

**Pull request merged**
